### ISSUE TITLE
How to make TFLM support Repeat/Tile op

### ISSUE THREAD

**Issue by AngLi666** [tensorflow/tflite-micro#2231]
I use tf.repeat in my model definition, and it will be converted into Tile op type after tf.lite.TFLiteConverter.
However, Tile is not include in TFLM, therefore, I could not compile my program correctly. Can someone help give a solution to resolve this issue? 
Is there a way to make TFLM support Tile?
Is there a way to workaround by removing/replacing Repeat/Tile ？

`
**
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

class SubSpectralNorm(layers.Layer):
    def __init__(self, channels, sub_bands, epsilon=1e-5):
        super(SubSpectralNorm, self).__init__()
        self.sub_bands = sub_bands
        self.bn = layers.BatchNormalization(axis=1, epsilon=epsilon)

    def call(self, x):
        N, F, T, C = x.shape
        x = tf.reshape(x, (N, F // self.sub_bands, T, C * self.sub_bands))
        x = tf.transpose(x, (0, 3, 1, 2))
        x = self.bn(x)
        x = tf.transpose(x, (0, 2, 3, 1))
        x = tf.reshape(x, (N, F, T, C))
        return x

class NormalBlock(tf.keras.Model):
    def __init__(self, n_chan, dilation=1, dropout=0.5, use_subspectral=True):
        super(NormalBlock, self).__init__()
        norm_layer = SubSpectralNorm(n_chan, 5) if use_subspectral else layers.BatchNormalization(axis=-1)
        self.f2 = tf.keras.Sequential([
            layers.Conv2D(n_chan, kernel_size=(3, 1), padding="same", groups=n_chan),
            norm_layer,
        ])
        self.f1 = tf.keras.Sequential([
            layers.Conv2D(n_chan, kernel_size=(1, 3), padding="same", groups=n_chan, dilation_rate=(1, dilation)),
            layers.BatchNormalization(axis=-1),
            layers.Activation(tf.nn.silu),
            layers.Conv2D(n_chan, kernel_size=1),
            layers.Dropout(dropout)
        ])
        self.activation = tf.keras.layers.ReLU()

    def call(self, x):
        n_freq = x.shape[2]
        x1 = self.f2(x)

        x2 = tf.reduce_mean(x1, axis=2, keepdims=True)
        x2 = self.f1(x2)
        x2 = tf.repeat(x2, repeats=n_freq, axis=2)

        return self.activation(x + x1 + x2)

class TransitionBlock(tf.keras.Model):
    def __init__(self, in_chan, out_chan, dilation=1, stride=1, dropout=0.5, use_subspectral=True):
        super(TransitionBlock, self).__init__()

        if stride == 1:
            conv = layers.Conv2D(out_chan, kernel_size=(3, 1), groups=out_chan, padding="same")
        else:
            conv = layers.Conv2D(out_chan, kernel_size=(3, 1), strides=(stride, 1), groups=out_chan, padding="same")

        norm_layer = SubSpectralNorm(out_chan, 5) if use_subspectral else layers.BatchNormalization(axis=-1)
        self.f2 = tf.keras.Sequential([
            layers.Conv2D(out_chan, kernel_size=(1, 1)),
            layers.BatchNormalization(axis=-1),
            layers.ReLU(),
            conv,
            norm_layer,
        ])
        
        self.f1 = tf.keras.Sequential([
            layers.Conv2D(out_chan, kernel_size=(1, 3), padding="same", groups=out_chan, dilation_rate=(1, dilation)),
            layers.BatchNormalization(axis=-1),
            layers.Activation(tf.nn.silu),
            layers.Conv2D(out_chan, kernel_size=1),
            layers.Dropout(dropout)
        ])

        self.activation = layers.ReLU()

    def call(self, x):
        x = self.f2(x)
        n_freq = x.shape[2]
        x1 = tf.reduce_mean(x, axis=2, keepdims=True)
        x1 = self.f1(x1)
        x1 = tf.repeat(x1, repeats=n_freq, axis=2)

        return self.activation(x + x1)

class BcResNet(tf.keras.Model):
    def __init__(self, n_class=35, scale=1, dropout=0.5, use_subspectral=True):
        super(BcResNet, self).__init__()

        self.input_conv = layers.Conv2D(16*scale, kernel_size=(5, 5), strides=(2, 1), padding="same")

        self.t1 = TransitionBlock(16*scale, 8*scale, dropout=dropout, use_subspectral=use_subspectral)
        self.n11 = NormalBlock(8*scale, dropout=dropout, use_subspectral=use_subspectral)

        self.t2 = TransitionBlock(8*scale, 12*scale, dilation=2, stride=2, dropout=dropout, use_subspectral=use_subspectral)
        self.n21 = NormalBlock(12*scale, dilation=2, dropout=dropout, use_subspectral=use_subspectral) 

        self.t3 = TransitionBlock(12*scale, 16*scale, dilation=4, stride=2, dropout=dropout, use_subspectral=use_subspectral)
        self.n31 = NormalBlock(16*scale, dilation=4, dropout=dropout, use_subspectral=use_subspectral)
        self.n32 = NormalBlock(16*scale, dilation=4, dropout=dropout, use_subspectral=use_subspectral)
        self.n33 = NormalBlock(16*scale, dilation=4, dropout=dropout, use_subspectral=use_subspectral)

        self.t4 = TransitionBlock(16*scale, 20*scale, dilation=8, dropout=dropout, use_subspectral=use_subspectral)
        self.n41 = NormalBlock(20*scale, dilation=8, dropout=dropout, use_subspectral=use_subspectral)
        self.n42 = NormalBlock(20*scale, dilation=8, dropout=dropout, use_subspectral=use_subspectral)
        self.n43 = NormalBlock(20*scale, dilation=8, dropout=dropout, use_subspectral=use_subspectral)

        self.dw_conv = layers.Conv2D(20*scale, kernel_size=(5, 5), groups=20)
        self.onexone_conv = layers.Conv2D(32*scale, kernel_size=1)

        self.head_conv = layers.Conv2D(n_class, kernel_size=1)
    
    def call(self, x):
        x = self.input_conv(x)
        x = self.t1(x)
        x = self.n11(x)

        x = self.t2(x)
        x = self.n21(x)

        x = self.t3(x)
        x = self.n31(x)
        x = self.n32(x)
        x = self.n33(x)

        x = self.t4(x)
        x = self.n41(x)
        x = self.n42(x)
        x = self.n43(x)

        x = self.dw_conv(x)
        x = self.onexone_conv(x)

        x = tf.reduce_mean(x, axis=2, keepdims=True)
        x = self.head_conv(x)

        x = tf.squeeze(x)

        return tf.nn.log_softmax(x, axis=-1)

    # get_concrete_function
    

import torch
import get_data
import numpy as np
def representative_dataset_gen():
    p = "D:\\datasets\\google_speech_command\\speech_commands_v0.02"
    calibrate_loader = torch.utils.data.DataLoader(
        get_data.SubsetSC(subset="train", path=p),
        batch_size=1,
        shuffle=True,
        drop_last=False,
        collate_fn=get_data.collate_fn,
        num_workers=0,
        pin_memory=True
    )

    cnt = 0
    for i_b, (data, _) in enumerate(calibrate_loader):
        if data.shape[2] != 40 or data.shape[3] != 32:
            continue
        if cnt >= 128:
            break
        data = data.numpy().transpose(0, 2, 3, 1).astype(np.float32)
        cnt += 1
        yield [data]
        
        

if __name__ == "__main__":
    net_name = "BcResNet"
    tflite_name = "./%s.tflite" % (net_name)
    # init BC-ResNet model
    model = BcResNet(scale=2)
    
    inp = np.ones(shape=(1, 40, 32, 1), dtype=np.float32)
    out = model(inp)

    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
    converter.target_spec.supported_types = [tf.int8]
    converter.inference_input_type = tf.int8
    converter.inference_output_type = tf.int8
    converter.representative_dataset = representative_dataset_gen
    model_tflite = converter.convert()
    
    open(tflite_name, "wb").write(model_tflite)
    print("tflite done")

**
`

<img width="1920" alt="20230919111513" src="https://github.com/tensorflow/tflite-micro/assets/77546516/818f0f98-760d-4c85-a38f-94c65a7d10fe">

**Comment by Dimfred** [comment]
Hello,
I am running into the same issue here.
Is there a workaround for this at least?

Code to reproduce:

```python
from pathlib import Path
import keras
import numpy as np
import tensorflow as tf
from keras.layers import LSTM, Dense, RepeatVector, TimeDistributed
from keras.models import Sequential

class Model(keras.Model):
    def __init__(self):
        super().__init__()

        self.model = Sequential()
        self.model.add(
            LSTM(
                32,
                activation="relu",
                input_shape=(n_timesteps, n_features),
                return_sequences=True,
            )
        )
        self.model.add(LSTM(16, activation="relu", return_sequences=False))
        self.model.add(RepeatVector(n_timesteps))
        self.model.add(LSTM(16, activation="relu", return_sequences=True))
        self.model.add(LSTM(32, activation="relu", return_sequences=True))
        self.model.add(TimeDistributed(Dense(n_features)))
        self.model.compile(optimizer="adam", loss="mse")
        self.model.build()
        self.model.summary()

    def call(self, inputs):
        return self.model(inputs)

model = Model()
model.predict(tf.random.normal((1, n_timesteps, n_features)))

concrete_func = run_model.get_concrete_function(
    tf.TensorSpec([1, n_timesteps, n_features], "float")
)
model.save("keras.tf", save_format="tf", signatures=concrete_func)

converter = tf.lite.TFLiteConverter.from_saved_model("keras.tf")
converter.target_spec.supported_types = [tf.float16]
converter.inference_input_type = tf.float32
converter.inference_output_type = tf.float32

tflite_model = converter.convert()
tflite_model_path = Path("keras.tflite")
tflite_model_path.write_bytes(bytes(tflite_model))
```

Would be happy about any suggestions.

**Comment by AngLi666** [comment]
It is dead way to work with TFLM.
You can try TVM if appliable to you

**Comment by AIWintermuteAI** [comment]
@Dimfred @AngLi666 Edge Impulse fork of tflite micro has Tile OP ported.
Disclaimer: I work for Edge Impulse and did the porting :)

**Comment by fiti-tsunche** [comment]
Does Edge Impulse open source your fork of tflite micro?

**Comment by AIWintermuteAI** [comment]
It's open source, you can find the code inside edge-impulse-sdk folder in C++ deployment - any project will do, the sdk is the same.
But for the question you asked - I think there are cases when TILE works with fixed output size and only those can be supported by tflite micro, as tflite micro does not support dynamic tensor sizes (or at least it didn't last time I checked - [here](https://github.com/tensorflow/tflite-micro/blob/9d35a74b03fa313026982abb4d2c4ffe29de88bc/tensorflow/lite/micro/kernels/batch_matmul_common.cc#L65) and some other comments in the code base)
What I did there was 
```
  if (IsDynamicTensor(output)) {
    TF_LITE_KERNEL_LOG(context, "Dynamic output tensor is not supported by tflite micro.");
    micro_context->DeallocateTempTfLiteTensor(input);
    micro_context->DeallocateTempTfLiteTensor(output);
    micro_context->DeallocateTempTfLiteTensor(multipliers);
    return kTfLiteError;
  }
```

**Comment by fiti-tsunche** [comment]
@AIWintermuteAI Thanks for your reply!

It seems I have to deeply understand tflite micro's memory management policy to make it support dynamic tensors. It would be difficult and dangerous.
